src/hotspot/os/linux/compilerThreadTimeout_linux.cpp

```diff
@@ -0,0 +1,128 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "compiler/compilerThread.hpp"
+#include "compiler/compileTask.hpp"
+#include "compilerThreadTimeout_linux.hpp"
+#include "oops/method.hpp"
+#include "runtime/osThread.hpp"
+#include "signals_posix.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+#include <pthread.h>
+
+#ifdef ASSERT
+void compiler_signal_handler(int signo, siginfo_t* info, void* context) {
+  CompilerThread::current()->timeout()->compiler_signal_handler(signo, info, context);
+}
+
+void CompilerThreadTimeoutLinux::compiler_signal_handler(int signo, siginfo_t* info, void* context) {
+  switch (signo) {
+    case TIMEOUT_SIGNAL: {
+      CompileTask* task = CompilerThread::current()->task();
+      const int SIZE = 512;
+      char method_name_buf[SIZE];
+      task->method()->name_and_sig_as_C_string(method_name_buf, SIZE);
+      assert(false, "compile task %d (%s) timed out after %zd ms",
+             task->compile_id(), method_name_buf, CompileTaskTimeout);
+    }
+    default: {
+      assert(false, "unexpected signal %d", signo);
+    }
+  }
+}
+#endif // ASSERT
+
+void CompilerThreadTimeoutLinux::arm() {
+#ifdef ASSERT
+  if (CompileTaskTimeout == 0) {
+    return;
+  }
+
+  const intx sec = (CompileTaskTimeout * NANOSECS_PER_MILLISEC) / NANOSECS_PER_SEC;
+  const intx nsec = (CompileTaskTimeout * NANOSECS_PER_MILLISEC) % NANOSECS_PER_SEC;
+  const struct timespec ts = {.tv_sec = sec, .tv_nsec = nsec};
+  const struct itimerspec its {.it_interval = ts, .it_value = ts};
+
+  // Start the timer.
+  timer_settime(_timer, 0, &its, nullptr);
+#endif // ASSERT
+}
+
+void CompilerThreadTimeoutLinux::disarm() {
+#ifdef ASSERT
+  if (CompileTaskTimeout == 0) {
+    return;
+  }
+
+  // Reset the timer by setting it to zero.
+  const struct itimerspec its {
+    .it_interval = {.tv_sec = 0, .tv_nsec=0},
+    .it_value = {.tv_sec = 0, .tv_nsec=0}
+  };
+  timer_settime(_timer, 0, &its, nullptr);
+#endif // ASSERT
+}
+
+bool CompilerThreadTimeoutLinux::init_timeout() {
+#ifdef ASSERT
+  if (CompileTaskTimeout == 0) {
+    return true;
+  }
+
+  JavaThread* thread = JavaThread::current();
+
+  // Create a POSIX timer sending SIGALRM to this thread only.
+  sigevent_t sev;
+  sev.sigev_value.sival_ptr = nullptr;
+  sev.sigev_signo = TIMEOUT_SIGNAL;
+  sev.sigev_notify = SIGEV_THREAD_ID;
+  #ifdef MUSL_LIBC
+  sev.sigev_notify_thread_id = thread->osthread()->thread_id();
+  #else
+  sev._sigev_un._tid = thread->osthread()->thread_id();
+  #endif // MUSL_LIBC
```

Comment:
The `ifdef` should probably also be without indentation like the other `ifdefs`:
```suggestion
#ifdef MUSL_LIBC
  sev.sigev_notify_thread_id = thread->osthread()->thread_id();
#else
  sev._sigev_un._tid = thread->osthread()->thread_id();
#endif // MUSL_LIBC
```

Reply:
Good catch.